### PR TITLE
RzBin: improve handling of addresses in ELF

### LINKED ISSUE #42: Tons of flags are shown at offset 0x00
```
$ rizin /bin/ls
[0x00000000]> s 0
[0x00000000]> pd 1
            ;-- section.:
            ;-- section..comment:
            ;-- section..shstrtab:
            ;-- segment.LOAD0:
            ;-- segment.GNU_STACK:
            ;-- segment.ehdr:
            ;-- __ctype_toupper_loc:
            ;-- getenv:
            ;-- cap_to_text:
            ;-- sigprocmask:
            ;-- __snprintf_chk:
            ;-- raise:
            ;-- free:
            ;-- abort:
            ;-- __errno_location:
            ;-- strncmp:
            ;-- _ITM_deregisterTMCloneTable:
            ;-- localtime_r:
            ;-- _exit:
            ;-- strcpy:
            ;-- __fpending:
            ;-- isatty:
            ;-- sigaction:
            ;-- iswcntrl:
            ;-- wcswidth:
            ;-- localeconv:
            ;-- mbstowcs:
            ;-- readlink:
            ;-- clock_gettime:
            ;-- setenv:
            ;-- textdomain:
            ;-- fclose:
            ;-- opendir:
            ;-- getpwuid:
            ;-- bindtextdomain:
            ;-- dcgettext:
            ;-- __ctype_get_mb_cur_max:
            ;-- strlen:
            ;-- __lxstat:
            ;-- __stack_chk_fail:
            ;-- getopt_long:
            ;-- mbrtowc:
            ;-- strchr:
            ;-- getgrgid:
            ;-- __fxstatat:
            ;-- snprintf:
            ;-- __overflow:
            ;-- strrchr:
            ;-- gmtime_r:
            ;-- lseek:
            ;-- __assert_fail:
            ;-- __strtoul_internal:
            ;-- fnmatch:
            ;-- memset:
            ;-- ioctl:
            ;-- getcwd:
            ;-- strspn:
            ;-- closedir:
            ;-- __libc_start_main:
            ;-- memcmp:
            ;-- _setjmp:
            ;-- fputs_unlocked:
            ;-- calloc:
            ;-- strcmp:
            ;-- signal:
            ;-- dirfd:
            ;-- fputc_unlocked:
            ;-- getpwnam:
            ;-- __memcpy_chk:
            ;-- syscall:
            ;-- sigemptyset:
            ;-- __gmon_start__:
            ;-- memcpy:
            ;-- getgrnam:
            ;-- tzset:
            ;-- fileno:
            ;-- tcgetpgrp:
            ;-- __xstat:
            ;-- readdir:
            ;-- wcwidth:
            ;-- malloc:
            ;-- fflush:
            ;-- nl_langinfo:
            ;-- strcoll:
            ;-- mktime:
            ;-- __freading:
            ;-- fwrite_unlocked:
            ;-- realloc:
            ;-- stpncpy:
            ;-- setlocale:
            ;-- __printf_chk:
            ;-- statx:
            ;-- timegm:
            ;-- strftime:
            ;-- mempcpy:
            ;-- memmove:
            ;-- error:
            ;-- fseeko:
            ;-- cap_get_file:
            ;-- unsetenv:
            ;-- cap_free:
            ;-- wcstombs:
            ;-- getxattr:
            ;-- gethostname:
            ;-- sigismember:
            ;-- exit:
            ;-- fwrite:
            ;-- __fprintf_chk:
            ;-- _ITM_registerTMCloneTable:
            ;-- fflush_unlocked:
            ;-- mbsinit:
            ;-- iswprint:
            ;-- __cxa_finalize:
            ;-- sigaddset:
            ;-- __ctype_tolower_loc:
            ;-- __ctype_b_loc:
            ;-- __sprintf_chk:
            ;-- rax:
            ;-- rbx:
            ;-- rcx:
            ;-- rdx:
            ;-- rsi:
            ;-- rdi:
            ;-- r8:
            ;-- r9:
            ;-- r10:
            ;-- r11:
            ;-- r12:
            ;-- r13:
            ;-- r14:
            ;-- r15:
            ;-- rflags:
            0x00000000      7f45           jg 0x47                     ; [26] ---- section size 247 named .shstrtab
```

### PR DESCRIPTION
**Pull request by ret2libc**
**Your checklist for this pull request**
- [x] I've read the [guidelines for contributing](https://github.com/rizinorg/rizin/blob/master/DEVELOPERS.md) to this repository
- [x] I made sure to follow the project's [coding style](https://github.com/rizinorg/rizin/blob/master/DEVELOPERS.md#code-style)
- [ ] I've documented or updated the documentation of every function and struct this PR changes. If not so I've explained why.
- [x] I've added tests that prove my fix is effective or that my feature works (if possible)
- [ ] I've updated the [rizin book](https://github.com/rizinorg/book) with the relevant information (if needed)

**Detailed description**

Improve handling of addresses in ELF to make clear when objects are not mapped in virtual space or not available in the file space. Also, it removes some objects that were mapped at address 0.

Before:
```
$ ./build-dev/binrz/rz-bin/rz-bin -i ./test/bins/elf/hello_world                                                                                                 [70/70]
[Imports]                              
nth vaddr      bind   type   lib name  
-------------------------------------------------------------
1   0x00000630 GLOBAL FUNC       free          
2   0x00000000 WEAK   NOTYPE     _ITM_deregisterTMCloneTable
3   0x00000640 GLOBAL FUNC       strcpy                                                                                                                                                                           
4   0x00000650 GLOBAL FUNC       puts                                                                                                                                                                             
5   0x00000660 GLOBAL FUNC       strlen                                                                                                                                                                           
6   0x00000000 GLOBAL FUNC       __libc_start_main                                                                                                                                                                
7   0x00000000 WEAK   NOTYPE     __gmon_start__
8   0x00000670 GLOBAL FUNC       malloc              
9   0x00000680 GLOBAL FUNC       strcat                                                 
10  0x00000000 WEAK   NOTYPE     _ITM_registerTMCloneTable
11  0x00000690 WEAK   FUNC       __cxa_finalize               
```
```
$ ./build-dev/binrz/rz-bin/rz-bin -S ./test/bins/elf/hello_world 2>/dev/null 
[Sections]
paddr      size  vaddr      vsize align perm name               type       flags         
-----------------------------------------------------------------------------------------
0x00000000 0x0   0x00000000 0x0   0x0   ----                    NULL       
0x00000238 0x1c  0x00000238 0x1c  0x0   -r-- .interp            PROGBITS   alloc
0x00000254 0x20  0x00000254 0x20  0x0   -r-- .note.ABI-tag      NOTE       alloc
0x00000274 0x24  0x00000274 0x24  0x0   -r-- .note.gnu.build-id NOTE       alloc
0x00000298 0x1c  0x00000298 0x1c  0x0   -r-- .gnu.hash          GNU_HASH   alloc
0x000002b8 0x120 0x000002b8 0x120 0x0   -r-- .dynsym            DYNSYM     alloc
0x000003d8 0xa3  0x000003d8 0xa3  0x0   -r-- .dynstr            STRTAB     alloc
0x0000047c 0x18  0x0000047c 0x18  0x0   -r-- .gnu.version       VERSYM     alloc
0x00000498 0x20  0x00000498 0x20  0x0   -r-- .gnu.version_r     VERNEED    alloc
0x000004b8 0xc0  0x000004b8 0xc0  0x0   -r-- .rela.dyn          RELA       alloc
0x00000578 0x90  0x00000578 0x90  0x0   -r-- .rela.plt          RELA       alloc,info
0x00000608 0x17  0x00000608 0x17  0x0   -r-x .init              PROGBITS   alloc,execute
0x00000620 0x70  0x00000620 0x70  0x0   -r-x .plt               PROGBITS   alloc,execute
0x00000690 0x8   0x00000690 0x8   0x0   -r-x .plt.got           PROGBITS   alloc,execute
0x000006a0 0x222 0x000006a0 0x222 0x0   -r-x .text              PROGBITS   alloc,execute
0x000008c4 0x9   0x000008c4 0x9   0x0   -r-x .fini              PROGBITS   alloc,execute
0x000008d0 0x14  0x000008d0 0x14  0x0   -r-- .rodata            PROGBITS   alloc
0x000008e4 0x3c  0x000008e4 0x3c  0x0   -r-- .eh_frame_hdr      PROGBITS   alloc
0x00000920 0x108 0x00000920 0x108 0x0   -r-- .eh_frame          PROGBITS   alloc
0x00000d90 0x8   0x00200d90 0x8   0x0   -rw- .init_array        INIT_ARRAY write,alloc
0x00000d98 0x8   0x00200d98 0x8   0x0   -rw- .fini_array        FINI_ARRAY write,alloc
0x00000da0 0x1f0 0x00200da0 0x1f0 0x0   -rw- .dynamic           DYNAMIC    write,alloc
0x00000f90 0x70  0x00200f90 0x70  0x0   -rw- .got               PROGBITS   write,alloc
0x00001000 0x10  0x00201000 0x10  0x0   -rw- .data              PROGBITS   write,alloc
0x00001010 0x0   0x00201010 0x8   0x0   -rw- .bss               NOBITS     write,alloc
0x00001010 0x25  0x00000000 0x25  0x0   ---- .comment           PROGBITS   merge,strings
0x00001038 0x660 0x00000000 0x660 0x0   ---- .symtab            SYMTAB     
0x00001698 0x26b 0x00000000 0x26b 0x0   ---- .strtab            STRTAB     
0x00001903 0xfe  0x00000000 0xfe  0x0   ---- .shstrtab          STRTAB     
```
```
$ ./build-dev/binrz/rizin/rizin -qc "pd 1 @ 0" ./test/bins/elf/hello_world 2>/dev/null 
            ;-- section.:
            ;-- section..comment:
            ;-- section..symtab:
            ;-- section..strtab:
            ;-- section..shstrtab:
            ;-- segment.LOAD0:
            ;-- segment.GNU_STACK:
            ;-- segment.ehdr:
            ;-- .comment:
            ;-- _ITM_deregisterTMCloneTable:
            ;-- __libc_start_main:
            ;-- __gmon_start__:
            ;-- _ITM_registerTMCloneTable:
        ┌─< 0x00000000      jg    0x47                                 ; [28] ---- section size 254 named .shstrtab
```

After:
```
$ ./build/binrz/rz-bin/rz-bin -i ./test/bins/elf/hello_world 2>/dev/null 
[Imports]
nth vaddr      bind   type   lib name                        
-------------------------------------------------------------
1   0x00000630 GLOBAL FUNC       free
2   ---------- WEAK   NOTYPE     _ITM_deregisterTMCloneTable
3   0x00000640 GLOBAL FUNC       strcpy
4   0x00000650 GLOBAL FUNC       puts
5   0x00000660 GLOBAL FUNC       strlen
6   ---------- GLOBAL FUNC       __libc_start_main
7   ---------- WEAK   NOTYPE     __gmon_start__
8   0x00000670 GLOBAL FUNC       malloc
9   0x00000680 GLOBAL FUNC       strcat
10  ---------- WEAK   NOTYPE     _ITM_registerTMCloneTable
11  0x00000690 WEAK   FUNC       __cxa_finalize
```
```
$ ./build/binrz/rz-bin/rz-bin -S ./test/bins/elf/hello_world 2>/dev/null                                                                                               
[Sections]
paddr      size  vaddr      vsize align perm name               type       flags
-----------------------------------------------------------------------------------------
0x00000000 0x0   ---------- 0x0   0x0   ----                    NULL
0x00000238 0x1c  0x00000238 0x1c  0x0   -r-- .interp            PROGBITS   alloc
0x00000254 0x20  0x00000254 0x20  0x0   -r-- .note.ABI-tag      NOTE       alloc
0x00000274 0x24  0x00000274 0x24  0x0   -r-- .note.gnu.build-id NOTE       alloc
0x00000298 0x1c  0x00000298 0x1c  0x0   -r-- .gnu.hash          GNU_HASH   alloc
0x000002b8 0x120 0x000002b8 0x120 0x0   -r-- .dynsym            DYNSYM     alloc
0x000003d8 0xa3  0x000003d8 0xa3  0x0   -r-- .dynstr            STRTAB     alloc
0x0000047c 0x18  0x0000047c 0x18  0x0   -r-- .gnu.version       VERSYM     alloc
0x00000498 0x20  0x00000498 0x20  0x0   -r-- .gnu.version_r     VERNEED    alloc
0x000004b8 0xc0  0x000004b8 0xc0  0x0   -r-- .rela.dyn          RELA       alloc
0x00000578 0x90  0x00000578 0x90  0x0   -r-- .rela.plt          RELA       alloc,info
0x00000608 0x17  0x00000608 0x17  0x0   -r-x .init              PROGBITS   alloc,execute
0x00000620 0x70  0x00000620 0x70  0x0   -r-x .plt               PROGBITS   alloc,execute
0x00000690 0x8   0x00000690 0x8   0x0   -r-x .plt.got           PROGBITS   alloc,execute
0x000006a0 0x222 0x000006a0 0x222 0x0   -r-x .text              PROGBITS   alloc,execute
0x000008c4 0x9   0x000008c4 0x9   0x0   -r-x .fini              PROGBITS   alloc,execute
0x000008d0 0x14  0x000008d0 0x14  0x0   -r-- .rodata            PROGBITS   alloc
0x000008e4 0x3c  0x000008e4 0x3c  0x0   -r-- .eh_frame_hdr      PROGBITS   alloc
0x00000920 0x108 0x00000920 0x108 0x0   -r-- .eh_frame          PROGBITS   alloc
0x00000d90 0x8   0x00200d90 0x8   0x0   -rw- .init_array        INIT_ARRAY write,alloc
0x00000d98 0x8   0x00200d98 0x8   0x0   -rw- .fini_array        FINI_ARRAY write,alloc
0x00000da0 0x1f0 0x00200da0 0x1f0 0x0   -rw- .dynamic           DYNAMIC    write,alloc
0x00000f90 0x70  0x00200f90 0x70  0x0   -rw- .got               PROGBITS   write,alloc
0x00001000 0x10  0x00201000 0x10  0x0   -rw- .data              PROGBITS   write,alloc
0x00001010 0x0   0x00201010 0x8   0x0   -rw- .bss               NOBITS     write,alloc
0x00001010 0x25  ---------- 0x25  0x0   ---- .comment           PROGBITS   merge,strings
0x00001038 0x660 ---------- 0x660 0x0   ---- .symtab            SYMTAB
0x00001698 0x26b ---------- 0x26b 0x0   ---- .strtab            STRTAB
0x00001903 0xfe  ---------- 0xfe  0x0   ---- .shstrtab          STRTAB
```
```
$ ./build/binrz/rizin/rizin -qc "pd 1 @ 0" ./test/bins/elf/hello_world 2>/dev/null 
            ;-- segment.LOAD0:
            ;-- segment.GNU_STACK:
            ;-- segment.ehdr:
            ;-- .comment:
        ┌─< 0x00000000      jg    0x47
```

**Test plan**

...

**Closing issues**

Closes https://github.com/rizinorg/rizin/issues/42

### CODE REVIEW

File: librz/bin/bin.c [hunk 1]
```diff
@@ -1239,19 +1239,17 @@ RZ_API RZ_OWN RzList *rz_bin_section_flag_to_list(RzBin *bin, ut64 flag) {
...
-	RzListIter *it, *it2;
+	RzListIter *it;
 	RzBinFile *bf;
-	RzBinSection *s;
 	rz_list_foreach (bin->binfiles, it, bf) {
-		// chk for baddr + size of no section is covering anything
-		// we should honor maps not sections imho
-		rz_list_foreach (bf->o->sections, it2, s) {
-			if (at >= s->vaddr && at < (s->vaddr + s->vsize)) {
+		if (bf->o) {
```
Comment (XVilka) [inline librz/bin/bin.c]: Reverse the condition and call `continue`?

File: librz/core/cbin.c [hunk 2]
```diff
@@ -1216,6 +1225,9 @@ RZ_API bool rz_core_bin_apply_relocs(RzCore *core, RzBinFile *binfile, bool va_b
 	char *sdb_module = NULL;
 	for (size_t i = 0; i < relocs->relocs_count; i++) {
 		RzBinReloc *reloc = relocs->relocs[i];
+		if ((va && reloc->vaddr == UT64_MAX) || (!va && reloc->paddr == UT64_MAX)) {
```
Comment (XVilka) [inline librz/core/cbin.c]: Maybe convert this pattern into macros?

**Review by XVilka — CHANGES_REQUESTED** [review]
Please add 'before' and 'after' in the PR description.

File: librz/bin/bin.c (hunk 1 above)
Comment (ret2libc) [reply librz/bin/bin.c]: i will, it is just wip for now as i wanted to see which tests failed

File: test/db/cmd/cmd_flags [hunk 3]
```diff
@@ -212,7 +212,7 @@ aaa
 flj~{254}
 EOF
 EXPECT=<<EOF
-{"name":"reloc.operator_delete_void","realname":"operator delete(void*)","size":8,"offset":16432}
+{"name":"obj.__dso_handle","realname":"__dso_handle","size":0,"offset":16496}
```
Comment (XVilka) [inline test/db/cmd/cmd_flags]: This is wrong - please change the index in `flj` to preserve the same symbol - please check what this test is about.

File: test/db/tools/rz_sign [hunk 4]
```diff
@@ -48,8 +48,8 @@ mkdir .tmp
 !rz-sign -o .tmp/hello_world.pat bins/elf/hello_world
 EOF
 EXPECT=<<EOF
-rz-sign: written 11 signatures to .tmp/hello_world.sig.
-rz-sign: written 11 signatures to .tmp/hello_world.pat.
+rz-sign: written 10 signatures to .tmp/hello_world.sig.
```
Comment (XVilka) [inline test/db/tools/rz_sign:51]: What symbol was removed here?

**Review by XVilka — CHANGES_REQUESTED** [review]
Very good UX change 💯 
Please fix the test first though.

File: test/db/tools/rz_sign (hunk 4 above)
Comment (ret2libc) [reply test/db/tools/rz_sign:51]: `loc.imp._ITM_deregisterTMCloneTable`, which was pointing at 0 and it was definitely not a function.

**Review by XVilka — APPROVED** [review]

**Comment by XVilka** [comment]
```

[XX] db/cmd/cmd_flags flj shows demangled symbols
RZ_NOPLUGINS=1 rizin -escr.utf8=0 -escr.color=0 -escr.interactive=0 -N -Qc 'aaa
flj~{240}
' bins/elf/demangle-test-cpp
-- stdout
--- expected
+++ actual
@@ -1,1 +1,1 @@
-{"name":"reloc.operator_delete_void","realname":"operator delete(void*)","size":8,"offset":16432}
+{"name":"reloc.std::__throw_length_error_char_const","realname":"std::__throw_length_error(char const*)","size":8,"offset":16424}

-- stderr
WARNING: Neither hash nor gnu_hash exist. Falling back to heuristics for deducing the number of dynamic symbols...
WARNING: Neither hash nor gnu_hash exist. Falling back to heuristics for deducing the number of dynamic symbols...
WARNING: Neither hash nor gnu_hash exist. Falling back to heuristics for deducing the number of dynamic symbols...
[ ] Analyze all flags starting with sym. and entry0 (aa)
[
[x] Analyze all flags starting with sym. and entry0 (aa)
```

**Comment by ret2libc** [comment]
Let's use squash+merge and not rebase please. I did a mess with commits, so it would be useless.

**Pull request merged**